src/linux/namespaces.rs

```diff
@@ -0,0 +1,334 @@
+    let put_old = new_root.join(OLD_ROOT_DIR);
+
+    // Ensure new root exists.
+    if !new_root.exists() {
+        fs::create_dir(&new_root)?;
+    }
+
+    // Create C-friendly versions for our paths.
+    let new_root_c = CString::new(new_root.as_os_str().as_bytes()).unwrap();
+    let put_old_c = CString::new(put_old.as_os_str().as_bytes()).unwrap();
+
+    // Create bind mount for new root to allow pivot.
+    bind_mount(&new_root_c, &new_root_c, 0)?;
+
+    // Sort bind mounts by shortest length, to create parents before their children.
+    let mut bind_mounts = bind_mounts.into_iter().collect::<Vec<_>>();
+    bind_mounts.sort_unstable_by(|(a_path, a_flags), (b_path, b_flags)| {
+        match a_path.components().count().cmp(&b_path.components().count()) {
+            Ordering::Equal => (a_path, a_flags).cmp(&(b_path, b_flags)),
+            ord => ord,
+        }
+    });
+
+    // Bind mount all allowed directories.
+    let current_dir = env::current_dir().ok();
+    for (mut path, flags) in bind_mounts {
+        // Ensure all paths are relative.
+        if path.is_relative() {
+            let current_dir = match &current_dir {
+                Some(current_dir) => current_dir,
+                // Ignore relative paths if we cannot access the working directory.
+                None => continue,
+            };
+            path = current_dir.join(path);
+        }
+
+        let src_c = CString::new(path.as_os_str().as_bytes()).unwrap();
+
+        // Get bind mount destination.
+        let unrooted_path = path.strip_prefix("/").unwrap();
+        let dst = new_root.join(unrooted_path);
+        let dst_c = CString::new(dst.as_os_str().as_bytes()).unwrap();
+
+        // Create mount target.
+        copy_tree(path, &new_root)?;
+
+        // Bind path with full permissions.
+        bind_mount(&src_c, &dst_c, flags)?;
+    }
+
+    // Bind mount old procfs.
+    let old_proc_c = CString::new("/proc").unwrap();
+    let new_proc = new_root.join("proc");
+    let new_proc_c = CString::new(new_proc.as_os_str().as_bytes()).unwrap();
+    fs::create_dir_all(&new_proc)?;
+    bind_mount(&old_proc_c, &new_proc_c, 0).unwrap();
+
+    // Pivot root to `new_root`, placing the old root in `put_old`.
+    fs::create_dir_all(put_old)?;
+    pivot_root(&new_root_c, &put_old_c)?;
+
+    // Remove the old root mount.
+    let new_old = PathBuf::from("/").join(OLD_ROOT_DIR);
+    let new_old_c = CString::new(new_old.as_os_str().as_bytes()).unwrap();
+    umount(&new_old_c)?;
+
+    // Prevent child mount namespaces from accessing this namespace's mounts.
+    deny_mount_propagation()?;
+
+    Ok(())
+}
+
+/// Replicate a directory tree under a different directory.
+///
+/// This will create all missing empty diretories and copy their permissions
+/// from the source tree.
+///
+/// If `src` ends in a file, an empty file with matching permissions will be
+/// created.
+fn copy_tree(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> Result<()> {
```

Comment:
While this copies the filesystem permissions, it does not copy the owner or group of the directories/files. In this scenario I don't think this is necessary, but probably a good idea to ponder this a little more.

Reply:
Most users and groups will be unmapped in the namespace at this point, right?

According to [`man 7 user_namespaces`](https://www.man7.org/linux/man-pages/man7/user_namespaces.7.html) (and consistent with my local testing), unmapped IDs are converted to the overflow UID (i.e., `nobody`).

So the only options for owner here are `root` or `nobody`.